File: packages/sdk/src/harness/memory/context_builder.py
"""
Context builder for assembling messages for LLM calls.

Handles system prompt injection, message windowing, and token budget management.
Includes automatic compression when context exceeds budget.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from pathlib import Path
from typing import TYPE_CHECKING, Any

from harness.memory.compressor import CompressionConfig, CompressionResult, ContextCompressor
from harness.memory.system_prompt import SystemPromptBuilder, SystemPromptConfig
from harness.memory.token_counter import TokenCounter
from harness.types import Message, Session

if TYPE_CHECKING:
    from harness.llm.base import ToolDefinition

logger = logging.getLogger(__name__)


@dataclass
class ContextBudget:
    """
    Token budget allocation for context components.

    Priority order: system_prompt > recent_messages > skills > memory

    Attributes:
        max_tokens: Maximum context window size
        response_reserve: Tokens reserved for LLM response
        system_prompt: Tokens allocated for system prompt
        tools: Tokens allocated for tool definitions
        recent_messages: Tokens allocated for recent messages
        skills: Tokens allocated for skill prompts
        memory: Tokens allocated for memory/context
    """
    max_tokens: int = 200000
    response_reserve: int = 4096
    system_prompt: int = 0
    tools: int = 0
    recent_messages: int = 0
    skills: int = 0
    memory: int = 0

    @property
    def available_for_input(self) -> int:
        """Tokens available for all input components."""
        return self.max_tokens - self.response_reserve

    @property
    def used(self) -> int:
        """Total tokens allocated."""
        return (
            self.system_prompt +
            self.tools +
            self.recent_messages +
            self.skills +
            self.memory
        )

    @property
    def remaining(self) -> int:
        """Tokens remaining unallocated."""
        return self.available_for_input - self.used

    @property
    def needs_compression(self) -> bool:
        """Check if context exceeds budget and needs compression."""
        return self.used > self.available_for_input

    @classmethod
    def allocate(
        cls,
        max_tokens: int,
        system_prompt_tokens: int = 0,
        tool_tokens: int = 0,
        message_ratio: float = 0.7,
        skills_ratio: float = 0.2,
        memory_ratio: float = 0.1,
    ) -> "ContextBudget":
        """
        Create a budget with automatic allocation.

        Args:
            max_tokens: Maximum context window
            system_prompt_tokens: Actual system prompt tokens
            tool_tokens: Actual tool definition tokens
            message_ratio: Ratio for messages (default 70%)
            skills_ratio: Ratio for skills (default 20%)
            memory_ratio: Ratio for memory (default 10%)

        Returns:
            Allocated ContextBudget
        """
        response_reserve = 4096
        available = max_tokens - response_reserve

        # Fixed allocations first (high priority)
        actual_system = min(system_prompt_tokens, available)
        remaining_after_system = available - actual_system

        actual_tools = min(tool_tokens, remaining_after_system)
        remaining = remaining_after_system - actual_tools

        # Proportional allocation for remaining
        total_ratio = message_ratio + skills_ratio + memory_ratio
        if total_ratio > 0:
            messages_alloc = int(remaining * message_ratio / total_ratio)
            skills_alloc = int(remaining * skills_ratio / total_ratio)
            memory_alloc = remaining - messages_alloc - skills_alloc
        else:
            messages_alloc = remaining
            skills_alloc = 0
            memory_alloc = 0

        return cls(
            max_tokens=max_tokens,
            response_reserve=response_reserve,
            system_prompt=actual_system,
            tools=actual_tools,
            recent_messages=messages_alloc,
            skills=skills_alloc,
            memory=memory_alloc,
        )


@dataclass
class ContextConfig:
    """Configuration for context building."""
    max_tokens: int = 200000
    system_prompt: str = ""
    window_size: int = 100  # Max number of recent messages
    compression_threshold: float = 0.9  # Compress when usage > 90%
    enable_compression: bool = True  # Enable automatic compression
    compression_config: CompressionConfig | None = None  # Compression settings

    # Dynamic system prompt configuration
    system_prompt_config: SystemPromptConfig | None = None  # Advanced prompt assembly
    project_root: Path | None = None  # Project root for AGENTS.md/MEMORY.md discovery
    memory_md_path: Path | None = None  # Optional path to global MEMORY.md file


@dataclass
class BuiltContext:
    """Result of context building."""
    messages: list[dict[str, Any]]
    system_prompt: str
    estimated_tokens: int
    budget: ContextBudget | None = None
    compression_needed: bool = False
    compression_result: CompressionResult | None = None


class ContextBuilder:
    """
    Builds context for LLM calls.

    Handles:
    - System prompt injection (static and dynamic)
    - Message windowing with token budget
    - Automatic compression detection and execution
    - Integration with TokenCounter for accurate counting
    - AGENTS.md / MEMORY.md discovery and loading
    """

    def __init__(
        self,
        config: ContextConfig | None = None,
        token_counter: TokenCounter | None = None,
        compressor: ContextCompressor | None = None,
    ):
        self.config = config or ContextConfig()
        self._token_counter = token_counter or TokenCounter()

        # Initialize compressor
        if compressor:
            self._compressor = compressor
        elif self.config.enable_compression:
            compression_config = self.config.compression_config or CompressionConfig()
            self._compressor = ContextCompressor(
                token_counter=self._token_counter,
                config=compression_config,
            )
        else:
            self._compressor = None

        # Initialize system prompt builder
        self._init_system_prompt_builder()

    def _init_system_prompt_builder(self) -> None:
        """Initialize the system prompt builder based on config."""
        logger.debug(
            f"_init_system_prompt_builder: system_prompt_config={self.config.system_prompt_config is not None}, "
            f"project_root={self.config.project_root}, system_prompt={bool(self.config.system_prompt)}, "
            f"memory_md_path={self.config.memory_md_path}"
        )
        if self.config.system_prompt_config:
            # Use provided config
            self._prompt_builder = SystemPromptBuilder(self.config.system_prompt_config)
        elif self.config.project_root or self.config.system_prompt or self.config.memory_md_path:
            # Create config from simple settings
            prompt_config = SystemPromptConfig(
                base_prompt=self.config.system_prompt,
                project_root=self.config.project_root,
                auto_discover=True,
            )
            self._prompt_builder = SystemPromptBuilder(prompt_config)

            # Add global memory source if specified
            if self.config.memory_md_path:
                logger.info(f"Adding GlobalMemory source: {self.config.memory_md_path}")
                from harness.memory.system_prompt import SystemPromptSource
                self._prompt_builder.add_source(SystemPromptSource(
                    name="GlobalMemory",
                    priority=40,
                    file_path=self.config.memory_md_path,
                ))
            else:
                logger.debug("No memory_md_path specified, skipping GlobalMemory source")
        else:
            # No dynamic prompt building
            self._prompt_builder = None

    def build(
        self,
        session: Session,
        new_prompt: str | None = None,
        tools: list[ToolDefinition] | None = None,
    ) -> BuiltContext:
        """
        Build context for LLM call with budget management.

        Automatically compresses context if it exceeds budget.

        Args:
            session: Current session
            new_prompt: Optional new user prompt
            tools: Available tools for budget estimation

        Returns:
            BuiltContext: Prepared messages for LLM
        """
        # Calculate budget
        budget = self._calculate_budget(tools)

        # Get messages from session
        session_messages = session.messages.copy()

        # Apply sliding window
        windowed_messages = self._apply_window(session_messages)

        # Add new prompt if provided
        if new_prompt:
            windowed_messages.append(Message(role="user", content=new_prompt))

        # Estimate current tokens
        estimated = self._estimate_tokens(windowed_messages)

        # Check if compression is needed
        compression_needed = estimated > budget.available_for_input * self.config.compression_threshold
        compression_result = None

        if compression_needed and self._compressor:
            logger.info(
                f"Context compression needed: {estimated} tokens > "
                f"{int(budget.available_for_input * self.config.compression_threshold)} threshold"
            )

            # Perform compression
            target_tokens = int(budget.available_for_input * 0.7)  # Aim for 70% utilization
            compression_result = self._compressor.compress(
                messages=windowed_messages,
                target_tokens=target_tokens,
            )

            # Use compressed messages
            windowed_messages = compression_result.compressed_messages
            estimated = compression_result.tokens_after

            logger.info(
                f"Compression complete: {compression_result.tokens_before} -> "
                f"{compression_result.tokens_after} tokens "
                f"(saved {compression_result.compression_saved})"
            )

        # Convert to API format
        messages = [msg.to_api_format() for msg in windowed_messages]

        return BuiltContext(
            messages=messages,
            system_prompt=self._get_system_prompt(),
            estimated_tokens=estimated,
            budget=budget,
            compression_needed=compression_needed,
            compression_result=compression_result,
        )

    def _get_system_prompt(self) -> str:
        """Get the system prompt, using dynamic builder if available."""
        if self._prompt_builder:
            return self._prompt_builder.build()
        return self.config.system_prompt

    def _calculate_budget(self, tools: list[ToolDefinition] | None = None) -> ContextBudget:
        """Calculate token budget allocation."""
        system_tokens = self._token_counter.count(self._get_system_prompt())
        tool_tokens = self._token_counter.estimate_tool_overhead(tools or [])

        return ContextBudget.allocate(
            max_tokens=self.config.max_tokens,
            system_prompt_tokens=system_tokens,
            tool_tokens=tool_tokens,
        )

    def _apply_window(self, messages: list[Message]) -> list[Message]:
        """
        Apply sliding window to messages.

        Args:
            messages: All messages

        Returns:
            Messages within window size
        """
        if len(messages) <= self.config.window_size:
            return messages

        return messages[-self.config.window_size:]

    def _estimate_tokens(self, messages: list[Message]) -> int:
        """
        Estimate total tokens in messages.

        Args:
            messages: Messages to estimate

        Returns:
            Estimated token count
        """
        total = 0
        for msg in messages:
            content = msg.content if isinstance(msg.content, str) else ""
            total += self._token_counter.count(content)
            total += 4  # Message format overhead
        return total

    def _build_messages(
        self,
        session: Session,
        new_prompt: str | None,
        token_budget: int,
    ) -> list[dict[str, Any]]:
        """
        Build message list within token budget.

        Uses sliding window to fit messages within budget.
        """
        messages = []

        # Start with window size limit
        recent = session.messages[-self.config.window_size:]

        # Add messages from newest to oldest until budget exhausted
        current_tokens = 0
        included_messages = []

        for msg in reversed(recent):
            msg_tokens = self._token_counter.count(msg.content if isinstance(msg.content, str) else "")

            if current_tokens + msg_tokens <= token_budget:
                included_messages.insert(0, msg)
                current_tokens += msg_tokens
            else:
                # Budget exhausted, stop adding older messages
                break

        # Convert to API format
        for msg in included_messages:
            messages.append(msg.to_api_format())

        # Add new prompt if provided
        if new_prompt:
            messages.append({
                "role": "user",
                "content": new_prompt,
            })

        return messages

    def _dict_to_message(self, msg_dict: dict[str, Any]) -> Message:
        """Convert dict to Message for token counting."""
        return Message(
            role=msg_dict.get("role", "user"),
            content=msg_dict.get("content", ""),
        )

    def set_system_prompt(self, prompt: str) -> None:
        """Set the system prompt."""
        self.config.system_prompt = prompt
        # Update prompt builder if exists
        if self._prompt_builder:
            # Update base prompt in config
            self._prompt_builder.config.base_prompt = prompt
            # Re-setup sources to pick up the new base prompt
            self._prompt_builder._setup_default_sources()

    def set_project_root(self, project_root: Path) -> None:
        """
        Set the project root for AGENTS.md / MEMORY.md discovery.

        Args:
            project_root: Path to project root directory
        """
        self.config.project_root = project_root
        # Re-initialize prompt builder with new project root
        self._init_system_prompt_builder()

    def add_prompt_source(self, source: "SystemPromptSource") -> None:
        """
        Add a custom system prompt source.

        Args:
            source: SystemPromptSource to add
        """
        if self._prompt_builder is None:
            # Initialize builder if not exists
            self._init_system_prompt_builder()
        self._prompt_builder.add_source(source)

    def get_available_prompt_sources(self) -> list[str]:
        """Get list of prompt sources that have content."""
        if self._prompt_builder:
            return self._prompt_builder.get_available_sources()
        return ["base"] if self.config.system_prompt else []

    def estimate_tokens(self, content: str) -> int:
        """Estimate token count for content using tiktoken."""
        return self._token_counter.count(content)

    def get_message_window(self, session: Session, max_tokens: int) -> list[Message]:
        """
        Get messages that fit within token budget.

        Args:
            session: Current session
            max_tokens: Maximum tokens for messages

        Returns:
            List of messages fitting within budget
        """
        messages = []
        current_tokens = 0

        for msg in reversed(session.messages):
            msg_tokens = self._token_counter.count(
                msg.content if isinstance(msg.content, str) else ""
            )

            if current_tokens + msg_tokens <= max_tokens:
                messages.insert(0, msg)
                current_tokens += msg_tokens
            else:
                break

        return messages

File: packages/sdk-java/harness-sdk-memory/src/main/java/com/harness/memory/ContextBuilder.java
package com.harness.memory;

import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.harness.core.TokenCounter;
import com.harness.types.Message;
import com.harness.types.Session;

/**
 * Context builder for assembling messages for LLM calls.
 *
 * Handles system prompt injection, message windowing, and token budget management.
 * Includes automatic compression when context exceeds budget.
 *
 * Features:
 * - Dynamic system prompt assembly from multiple sources (AGENTS.md, MEMORY.md, etc.)
 * - Automatic compression when context exceeds budget
 * - Sliding window for message history
 * - Token budget allocation
 *
 * Example:
 * <pre>
 * ContextBuilder builder = new ContextBuilder()
 *     .withMaxTokens(200000)
 *     .withSystemPrompt("You are a helpful assistant.")
 *     .withProjectRoot(Path.of("/path/to/project"));
 *
 * BuiltContext context = builder.build(session);
 * String systemPrompt = context.systemPrompt();
 * List<Message> messages = context.messages();
 * </pre>
 */
public class ContextBuilder {

    private static final Logger logger = LoggerFactory.getLogger(ContextBuilder.class);

    private ContextConfig config;
    private final TokenCounter tokenCounter;
    private ContextCompressor compressor;
    private SystemPromptBuilder promptBuilder;

    public ContextBuilder() {
        this(ContextConfig.defaults());
    }

    public ContextBuilder(ContextConfig config) {
        this.config = config;
        this.tokenCounter = new TokenCounter();
        initCompressor();
        initSystemPromptBuilder();
    }

    /**
     * Initialize compressor based on config.
     */
    private void initCompressor() {
        if (config.enableCompression()) {
            CompressionConfig compressionConfig = config.compressionConfig() != null
                ? config.compressionConfig()
                : CompressionConfig.defaults();
            this.compressor = new ContextCompressor(tokenCounter, compressionConfig);
        } else {
            this.compressor = null;
        }
    }

    /**
     * Initialize system prompt builder based on config.
     */
    private void initSystemPromptBuilder() {
        if (config.systemPromptConfig() != null) {
            // Use provided config
            this.promptBuilder = new SystemPromptBuilder(config.systemPromptConfig());
        } else if (config.projectRoot() != null
                   || (config.systemPrompt() != null && !config.systemPrompt().isEmpty())
                   || config.memoryMdPath() != null) {
            // Create config from simple settings
            SystemPromptConfig promptConfig = SystemPromptConfig.builder()
                .basePrompt(config.systemPrompt())
                .projectRoot(config.projectRoot())
                .autoDiscover(true)
                .build();

            this.promptBuilder = new SystemPromptBuilder(promptConfig);

            // Add global memory source if specified
            if (config.memoryMdPath() != null) {
                Path memoryPath = config.memoryMdPath();
                // memoryMdPath can be either a directory or full file path
                if (memoryPath.toFile().isDirectory()) {
                    memoryPath = memoryPath.resolve("MEMORY.md");
                }
                this.promptBuilder.addSource(new SystemPromptSource(
                    "GlobalMemory",
                    40,
                    null,
                    memoryPath,
                    false
                ));
            }
        } else {
            // No dynamic prompt building
            this.promptBuilder = null;
        }
    }

    /**
     * Build context from session.
     *
     * @param session Current session
     * @return BuiltContext with messages and system prompt
     */
    public BuiltContext build(Session session) {
        return build(session, null, null);
    }

    /**
     * Build context from session with optional new prompt and tools.
     *
     * @param session    Current session
     * @param newPrompt  Optional new user prompt
     * @param tools      Available tools for budget estimation
     * @return BuiltContext with messages and system prompt
     */
    public BuiltContext build(Session session, String newPrompt, List<Object> tools) {
        // Build system prompt once (avoid duplicate build calls)
        String systemPrompt = getSystemPrompt();

        // Calculate budget using pre-built system prompt
        ContextBudget budget = calculateBudget(tools, systemPrompt);

        // Get messages from session
        List<Message> sessionMessages = new ArrayList<>(session.messages());

        // Apply sliding window
        List<Message> windowedMessages = applyWindow(sessionMessages);

        // Add new prompt if provided
        if (newPrompt != null && !newPrompt.isEmpty()) {
            windowedMessages.add(Message.user(newPrompt));
        }

        // Estimate current tokens
        int estimated = estimateTokens(windowedMessages);

        // Check if compression is needed
        int threshold = (int) (budget.availableForInput() * config.compressionThreshold());
        boolean compressionNeeded = estimated > threshold;
        CompressionResult compressionResult = null;

        if (compressionNeeded && compressor != null) {
            logger.info("Context compression needed: {} tokens > {} threshold",
                estimated, threshold);

            // Perform compression
            int targetTokens = (int) (budget.availableForInput() * 0.7);  // Aim for 70% utilization
            compressionResult = compressor.compress(windowedMessages, targetTokens);

            // Use compressed messages
            windowedMessages = new ArrayList<>(compressionResult.compressedMessages());
            estimated = compressionResult.tokensAfter();

            // If compression generated a summary, prepend it to system prompt
            // This ensures system message stays first (required by chat templates)
            if (compressionResult.summary() != null && !compressionResult.summary().isEmpty()) {
                systemPrompt = systemPrompt + "\n\n[Previous conversation summary]\n" + compressionResult.summary();
            }

            logger.info("Compression complete: {} -> {} tokens (saved {})",
                compressionResult.tokensBefore(), compressionResult.tokensAfter(),
                compressionResult.compressionSaved());
        }

        return new BuiltContext(
            windowedMessages,
            systemPrompt,
            estimated,
            budget,
            compressionNeeded,
            compressionResult
        );
    }

    /**
     * Get the system prompt, using dynamic builder if available.
     */
    private String getSystemPrompt() {
        if (promptBuilder != null) {
            return promptBuilder.build();
        }
        return config.systemPrompt() != null ? config.systemPrompt() : "";
    }

    /**
     * Calculate token budget allocation.
     *
     * @param tools          Available tools for budget estimation
     * @param systemPrompt   Pre-built system prompt (avoid duplicate build() calls)
     */
    private ContextBudget calculateBudget(List<Object> tools, String systemPrompt) {
        // Use provided systemPrompt to avoid duplicate build() calls
        if (systemPrompt == null) {
            systemPrompt = getSystemPrompt();
        }
        int systemTokens = tokenCounter.count(systemPrompt);
        int toolTokens = estimateToolOverhead(tools);

        return ContextBudget.allocate(
            config.maxTokens(),
            systemTokens,
            toolTokens
        );
    }

    /**
     * Estimate token overhead for tool definitions.
     *
     * Each tool definition adds overhead for:
     * - Tool name and description
     * - Input schema (JSON)
     * - Format overhead
     *
     * @param tools List of tools (can be Tool objects or ToolDefinition objects)
     * @return Estimated token overhead
     */
    private int estimateToolOverhead(List<Object> tools) {
        if (tools == null || tools.isEmpty()) {
            return 0;
        }

        int overhead = 0;
        for (Object tool : tools) {
            // Base overhead per tool: ~50 tokens for name, description, format
            overhead += 50;

            // If it's a Tool, estimate schema overhead
            if (tool instanceof com.harness.core.Tool) {
                com.harness.core.Tool t = (com.harness.core.Tool) tool;
                if (t.inputSchema() != null) {
                    // Rough estimate: JSON schema adds ~100-200 tokens
                    overhead += 150;
                }
            }
        }

        return overhead;
    }

    /**
     * Apply sliding window to messages.
     *
     * @param messages All messages
     * @return Messages within window size
     */
    private List<Message> applyWindow(List<Message> messages) {
        if (messages.size() <= config.windowSize()) {
            return messages;
        }
        return new ArrayList<>(messages.subList(messages.size() - config.windowSize(), messages.size()));
    }

    /**
     * Estimate total tokens in messages.
     *
     * @param messages Messages to estimate
     * @return Estimated token count
     */
    private int estimateTokens(List<Message> messages) {
        int total = 0;
        for (Message msg : messages) {
            String content = msg.contentAsString();
            total += tokenCounter.count(content);
            total += 4;  // Message format overhead
        }
        return total;
    }

    /**
     * Build message list within token budget.
     *
     * Uses sliding window to fit messages within budget.
     *
     * @param session      Current session
     * @param newPrompt    Optional new user prompt
     * @param tokenBudget  Maximum tokens for messages
     * @return List of messages in API format
     */
    private List<Message> buildMessages(Session session, String newPrompt, int tokenBudget) {
        List<Message> messages = new ArrayList<>();

        // Start with window size limit
        int windowSize = config.windowSize();
        List<Message> recent = session.messages().size() > windowSize
            ? new ArrayList<>(session.messages().subList(session.messages().size() - windowSize, session.messages().size()))
            : new ArrayList<>(session.messages());

        // Add messages from newest to oldest until budget exhausted
        int currentTokens = 0;
        List<Message> includedMessages = new ArrayList<>();

        for (int i = recent.size() - 1; i >= 0; i--) {
            Message msg = recent.get(i);
            int msgTokens = tokenCounter.count(msg.contentAsString());

            if (currentTokens + msgTokens <= tokenBudget) {
                includedMessages.add(0, msg);
                currentTokens += msgTokens;
            } else {
                // Budget exhausted, stop adding older messages
                break;
            }
        }

        messages.addAll(includedMessages);

        // Add new prompt if provided
        if (newPrompt != null && !newPrompt.isEmpty()) {
            messages.add(Message.user(newPrompt));
        }

        return messages;
    }

    /**
     * Get messages that fit within token budget.
     *
     * @param session    Current session
     * @param maxTokens  Maximum tokens for messages
     * @return List of messages fitting within budget
     */
    public List<Message> getMessageWindow(Session session, int maxTokens) {
        List<Message> messages = new ArrayList<>();
        int currentTokens = 0;

        for (int i = session.messages().size() - 1; i >= 0; i--) {
            Message msg = session.messages().get(i);
            int msgTokens = tokenCounter.count(msg.contentAsString());

            if (currentTokens + msgTokens <= maxTokens) {
                messages.add(0, msg);
                currentTokens += msgTokens;
            } else {
                break;
            }
        }

        return messages;
    }

    // Configuration methods (mutate config and re-initialize if needed)

    /**
     * Set the maximum context window tokens.
     */
    public ContextBuilder withMaxTokens(int maxTokens) {
        this.config.setMaxTokens(maxTokens);
        return this;
    }

    /**
     * Set the base system prompt.
     */
    public ContextBuilder withSystemPrompt(String systemPrompt) {
        this.config.setSystemPrompt(systemPrompt);
        // Update prompt builder if exists
        if (promptBuilder != null) {
            // Update base prompt in config by recreating builder
            initSystemPromptBuilder();
        }
        return this;
    }

    /**
     * Set the message window size.
     */
    public ContextBuilder withWindowSize(int windowSize) {
        this.config.setWindowSize(windowSize);
        return this;
    }

    /**
     * Enable or disable compression.
     */
    public ContextBuilder withCompressionEnabled(boolean enabled) {
        this.config.setEnableCompression(enabled);
        initCompressor();
        return this;
    }

    /**
     * Set the project root for AGENTS.md / MEMORY.md discovery.
     */
    public ContextBuilder withProjectRoot(Path projectRoot) {
        this.config.setProjectRoot(projectRoot);
        // Re-initialize prompt builder with new project root
        initSystemPromptBuilder();
        return this;
    }

    /**
     * Set the path to global MEMORY.md file.
     */
    public ContextBuilder withMemoryMdPath(Path memoryMdPath) {
        this.config.setMemoryMdPath(memoryMdPath);
        // Re-initialize prompt builder with new memory path
        initSystemPromptBuilder();
        return this;
    }

    /**
     * Set the system prompt directly.
     *
     * @param prompt The system prompt to use
     */
    public void setSystemPrompt(String prompt) {
        this.config.setSystemPrompt(prompt);
        // Update prompt builder if exists
        if (promptBuilder != null) {
            // Update base prompt in config by recreating the builder
            initSystemPromptBuilder();
        }
    }

    /**
     * Set the project root for AGENTS.md / MEMORY.md discovery.
     *
     * @param projectRoot Path to project root directory
     */
    public void setProjectRoot(Path projectRoot) {
        this.config.setProjectRoot(projectRoot);
        // Re-initialize prompt builder with new project root
        initSystemPromptBuilder();
    }

    /**
     * Add a custom system prompt source.
     *
     * @param source SystemPromptSource to add
     */
    public void addPromptSource(SystemPromptSource source) {
        if (promptBuilder == null) {
            // Initialize builder if not exists
            initSystemPromptBuilder();
        }
        if (promptBuilder != null) {
            promptBuilder.addSource(source);
        }
    }

    /**
     * Get list of prompt sources that have content.
     *
     * @return List of available source names
     */
    public List<String> getAvailablePromptSources() {
        if (promptBuilder != null) {
            return promptBuilder.getAvailableSources();
        }
        List<String> result = new ArrayList<>();
        if (config.systemPrompt() != null && !config.systemPrompt().isEmpty()) {
            result.add("base");
        }
        return result;
    }

    /**
     * Estimate token count for content using tiktoken.
     *
     * @param content Content to estimate
     * @return Estimated token count
     */
    public int estimateTokens(String content) {
        return tokenCounter.count(content);
    }
}